### ISSUE TITLE
API should perform basic permissions check on all requests

### ISSUE THREAD

**Issue by jmiranda** [annotationsatharvard/catcha#70]
Luis emailed the following enhancement on Aug 7: 
> Security/Permissions: I have a question and will be testing it to double check. Does the server do any sort of permission checking whenever a user tries to read/update/delete an annotation? For example, each annotation has a permissions ("read" = ['user_id1']). If it's empty it means anyone can read it, if it has at least one user_id, it means that list of users can read it. My question is: does/can the server check the "read" permission? So if I say give me all annotations for all users can it give me only those annotations where "read"=[] or where "read"=["user_making_the_request"]? Likewise if a user makes an update/delete request can it be checked whether they are permitted in the appropriate list (or the "admin" list)?

**Comment by lduarte1991** [comment]
What we would like CATCH to do is the following (currently the search query is the only one affected by this):

1. User sends a search request (their user_id is included in the token). 
2. The result is a list of annotations.
3. For each annotation, if the read permission is an empty array then yes return that array.
4. If the read permission is not empty, check to see if the user_id in the token is in that read list (If it's not empty, then 99% of the time it should only have 1 user_id—the creator's)
5. (optional step…need to discuss privacy expectations with Jon/Artie here)If their username is not in that list, we check to see if they are an admin. If they are in the admin list (I.e. An instructor trying to read all annotations) then it also gets returned.
6. If none of steps 3-5 are met then that annotation is not returned to the user who made the query.

**Comment by jmiranda** [comment]
@lduarte1991 So two issues 

1. Are you planning to pass me a parameter to let me know whether the user is an admin / instructor? If so, that should probably go into the token payload.

2. Given that the permission check is happening outside of the database there's no way for me to return the exact number of records you'd like (e.g. `limit` parameter). For example, let's say you send a Search API request where you want `limit=50` where `uri=abc`. The API will search for 50 annotations that match the URI and then we'll post-process those results to come up with annotations that the current user is allowed to view. In most cases this result will be less than 50. Are you ok with that?

**Comment by arthurian** [comment]
@jmiranda @lduarte1991  When you say the permission check is happening outside of the database and there's no way to return the exact number of records, I'm assuming the procedure is something like this:

1. Query DB for annotation records, limiting the number of rows by the `limit` and `offset` parameters.
2. Filter the query results to the subset that `user_id` is authorized to read by checking the permissions object contained in each record.
3. Return the results. In most cases with `total < `limit`.

As you mentioned in your email, the best approach would be if the `permissions` were stored in a separate table and could be filtered at the query level, but that would require data migration and more time. 

Couldn't the procedure above be modified to guarantee that *if* there are `limit` records that `user_id` is authorized to read, then `limit` records will be returned? Something like this perhaps?

1. Build an in-memory index of annotation records that the user is authorized to read.
    1. Fetch `id, permissions` for all annotation records matching the `uri` (one query with the required fields).
    2. Reduce the records to a list of `ids` that the user is authorized to read.
2. Fetch the annotation objects up to the limit.
    1. Slice the list of `ids` up to the `limit`.
    2. Fetch the annotations given by the slice of `ids`.

The result would be `limit` records, or if there's less than that, then that means there are no more records the user is authorized to read. 

Maybe (1) isn't feasible if the `count()` for that query exceeds a certain threshold (500? 1K? 10K? 25K?), in which case it would require some kind of pagination/iteration. Not ideal, obviously, just wondering if it's an option given the size of the datasets we're seeing in practice.

**Comment by lduarte1991** [comment]
It _technically_ isn't such a huge deal since in the latest iterations of the tool I've been asking for all of the annotations for a specific uri at once and my own system handles a pagination aspect (i.e. I ask for all public annotations and then my own library handles pagination client-side, so in reality they will get 50 if that's what they wanted). The following only applies if like you mentioned yesterday you would like us to set a particular limit to how many annotations I can call up:

I feel like this could cause some serious problems if we have to specifically set a limit (currently used for pagination purposes). If ```limit=50``` and 20 students come in and each create 1 personal annotation, then the next person could at most see 30 of any public annotations created.

**Comment by jmiranda** [comment]
Yes, that approach crossed my mind. The problem is that we'd need to maintain the index for every user / annotation in the system (there are around 250,000 annotations) . So instead of building our own in-memory index I was planning to use ElasticSearch or Solr (and doing the search through them).  But that's slightly more difficult than the "permission" table hack that I am proposing, so it might take a few more weeks.  Long term / ideal solution would be to ONLY store the JSON objects in MongoDB (or some other appropriate doc storage) rather than a RDBMS like MySQL. That way we'd be able to query on any attribute in the document (including permissions) and have a single repository for storage and querying. Not sure how fast MongoDB would be on large data sets so perhaps an ElasticSearch / Solr solution would be necessary in either case.

**Comment by jmiranda** [comment]
@lduarte1991 There's not a huge performance hit with the new code. I was able to get 500 - 1000 annotations within a second or two. But if you're pulling more than that then I assume we may have some issues.

**Comment by lduarte1991** [comment]
@jmiranda @arthurian I talked with Rashmi for a while and she had a suggestion. It's not the most elegant thing but it might be something that can be done much a bit faster? Currently the only reason that permissions blob exists is that it was part of the original annotator code. We were wondering if it would be a better idea to make each of the items in the permissions blob be their own index fields akin to how we added ```context_id``` and ```collection_id``` a while back. 

If we do that, then it would only take a little finagling in my code to query for particular items. If we were to add a ```read_permissions``` (and the others) as separate fields, I would then make sure that I send both the blob and the new fields in when a user creates an annotation. That way I can query for read permissions directly. It shouldn't be too difficult to set a default because ALL the annotations should currently have an empty array as their read. The other fields ```admin, delete, update``` would be a little harder because they all have to default to an array of the owner's user id but for now read is the one that matters most since I already do my own checking for creating, deleting and updating annotations.

**Comment by jmiranda** [comment]
Yes, that's essentially what I was referring to by the "permission table" hack. Whether it's a separate table or just a column for each permission on the annotation table is a design decision, but the same work (i.e. new tables/columns, data migration for existing annotations, storing permissions on update/create) will be required to get it to work correctly.

**Comment by jmiranda** [comment]
If you're saying that you don't care about migrating the existing annotations, then we can probably get this done by early next week. And it wouldn't require you to change anything on your end (at least, I don't think you do). I can just add the permission table, persist permissions, and modify the query.

**Comment by lduarte1991** [comment]
Yeah I don't really care about migrating existing ones cause technically everything is set to public.

**Comment by arthurian** [comment]
@jmiranda I can see why that approach won't work -- 250K annotations, yikes!

I agree with @lduarte1991, we don't need to migrate any of our existing annotations either since it's all set to public anyway. If we can get this done by early next week, that would work for us.

**Comment by jmiranda** [comment]
@lduarte1991 @arthurian Do we start to normalize user data while we're at it? I assume we'll just have username or email, so it'll be a pretty small table (id, username, email) but it will likely help with performance when we join the annotations / permissions / users in the query (rather than indexing on strings - i.e. username/email).

**Comment by arthurian** [comment]
@jmiranda That sounds like a good idea, while we're at it. I think all we'll need is `(id, user_id, name, email)`. So `user_id` instead of `username` since we're storing an anonymous ID sent by the LTI tool consumer (i.e. edX or Canvas) to identify the user. The `email` should be optional since I don't think we're currently storing that, but we might want to later.

@lduarte1991 Is that right?
